linera-execution/src/resources.rs

```diff
@@ -265,135 +274,90 @@ impl AsMut<ResourceTracker> for ResourceTracker {
-    C: Context + Clone + Send + Sync + 'static,
-    ViewError: From<C::Error>,
-{
-    fn as_amount(&self) -> Amount {
-        match &self.owner {
-            None => *self.view.system.balance.get(),
-            Some(owner) => self
-                .view
-                .system
-                .balance
-                .get()
-                .try_add(self.get_owner_balance(owner))
-                .expect("Overflow was tested in `ResourceController::with` and `add_assign`"),
+impl BalanceHolder for Sources<'_> {
+    fn balance(&self) -> Result<Amount, ArithmeticError> {
+        let mut amount = Amount::ZERO;
+        for source in self.sources.iter() {
+            amount.try_add_assign(**source)?;
         }
+        Ok(amount)
     }
 
     fn try_add_assign(&mut self, other: Amount) -> Result<(), ArithmeticError> {
-        match self.owner {
-            None => self.view.system.balance.get_mut().try_add_assign(other),
-            Some(owner) => {
-                // Try to credit the owner account first.
-                // TODO(#1648): This may need some additional design work.
-                let balance = self.get_owner_balance_mut(&owner);
-                balance.try_add_assign(other)?;
-                // Safety check. (See discussion below in `ResourceController::with`).
-                balance.try_add(*self.view.system.balance.get())?;
-                Ok(())
-            }
-        }
+        // Try to credit the owner account first.
+        // TODO(#1648): This may need some additional design work.
+        let source = self.sources.last_mut().expect("at least one source");
+        source.try_add_assign(other)
     }
 
-    fn try_sub_assign(&mut self, other: Amount) -> Result<(), ArithmeticError> {
-        match self.owner {
-            None => self.view.system.balance.get_mut().try_sub_assign(other),
-            Some(owner) => {
-                // Charge the chain account first then the owner's account.
-                if self
-                    .view
-                    .system
-                    .balance
-                    .get_mut()
-                    .try_sub_assign(other)
-                    .is_err()
-                {
-                    let balance = *self.view.system.balance.get();
-                    *self.view.system.balance.get_mut() = Amount::ZERO;
-                    let delta = other.try_sub(balance).expect("balance < other");
-                    self.get_owner_balance_mut(&owner).try_sub_assign(delta)?;
-                }
-                Ok(())
+    fn try_sub_assign(&mut self, mut other: Amount) -> Result<(), ArithmeticError> {
+        for source in self.sources.iter_mut() {
+            if source.try_sub_assign(other).is_ok() {
+                return Ok(());
             }
+            other.try_sub_assign(**source).expect("*source < other");
+            **source = Amount::ZERO;
+        }
+        if other > Amount::ZERO {
+            // TODO: Provide a better error.
```

Comment:
Should we create a ticket for that?

Reply:
Oops, I was supposed to get back to this.

Reply:
I did find unmapped underflows in system.rs though